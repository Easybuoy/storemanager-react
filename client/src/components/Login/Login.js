import React, { Component } from "react";
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { signIn } from '../../actions/authActions';
class Login extends Component {
    constructor() {
        super();
        this.state = {
            email: '',
            password: '',
        }

        this.onChange = this.onChange.bind(this);
        this.onSubmit = this.onSubmit.bind(this);

    }

    onChange(event){
        this.setState({[event.target.name] : event.target.value});
    }


    onSubmit(e) {
        e.preventDefault();
        const userData = {
            email: this.state.email,
            password: this.state.password 
        };

        this.props.signIn(userData);
    }

    render() { 
        const { errors } = this.props;

        if (this.props.auth.isSignedIn ) {
            this.props.history.push('/dashboard')
        }

        return (
            <div id="loginbody">
        <div className="loginbox">
            <img src="https://store--manager.herokuapp.com/assets/uploads/users/default-avatar.png" className="avatar"/>
            <h2 className="text-center">Login to access store manager</h2>
            <form>
                <p>Email</p>
                <input onChange={this.onChange} type="text" id="loginusername" name="email" placeholder="Enter Email" />
        <div className="login-error">{errors.email}</div>
                <p>Password</p>
                <input onChange= {this.onChange} type="password" id="loginpassword" name="password" placeholder="Enter Password" />
                <div className="login-error">{errors.password} <br /><br /></div>
                <input onClick={this.onSubmit} type="submit" name="submit" id="loginsubmit" value="Login" />
            </form>
        </div>

            </div>
        );
    }
}

Login.propTypes = {
    signIn: PropTypes.func.isRequired,
    auth: PropTypes.object.isRequired,
    errors: PropTypes.object.isRequired,
}

const mapStateToProps = (state) => ({
    auth: state.auth,
    errors: state.errors,
});

export default connect(mapStateToProps, { signIn })(Login);