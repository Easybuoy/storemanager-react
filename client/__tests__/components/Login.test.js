import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';

import { Login } from '../../src/components/Login/Login';

describe('<Login />', () => {
    const props = {
        signIn: jest.fn(),
        auth: jest.mock(),
        errors: jest.mock(),
        history: {push: jest.fn()}
      };

    it('renders the Login component correctly', () => {
        const wrapper = shallow(<Login {...props}/>);
        expect(wrapper).toMatchSnapshot();
    });

    it('should call the mock signIn function', () => {
        const wrapper = shallow(<Login {...props} />)
 
         wrapper.find('form').simulate(
           'submit', 
           {preventDefault() {}}
         )
         expect(props.signIn).toBeCalled();
         expect(props.signIn).toHaveReturned();
        });

    it('should call the mock onChange function', () => {
        const wrapper = shallow(<Login {...props} />)

        const event = {target: {name: "pollName", value: "spam"}};

         sinon.spy(wrapper.instance(), 'onChange');
         wrapper.instance().onChange(event);
         expect(wrapper.instance().onChange.calledOnce)
           .toEqual(true);
           expect(wrapper.instance().onChange.calledWith(event));

        });

        it('should call the mock onSubmit function', () => {
            const wrapper = shallow(<Login {...props} />)
    
            const event = Object.assign(jest.fn(), { preventDefault: () => {} });
    
             sinon.spy(wrapper.instance(), 'onSubmit');
             wrapper.instance().onSubmit(event);
             expect(wrapper.instance().onSubmit.calledOnce)
               .toEqual(true);
               expect(wrapper.instance().onSubmit.calledWith(event));
            });

            it('should show error message(s) if login error occurs', () => {
              props.errors = { message: 'Network Error', email: 'Email field is required', password: 'Password field is required'}
              const wrapper = shallow(<Login {...props} />)
  
              expect(props.signIn).toBeCalled();
               expect(props.signIn).toHaveReturned();

              });

              it('should redirrect if user is signed in', () => {
                props.auth.isSignedIn = true;
                const wrapper = shallow(<Login {...props} />)
    
                expect(props.signIn).toBeCalled();
                 expect(props.signIn).toHaveReturned();
  
                });
});