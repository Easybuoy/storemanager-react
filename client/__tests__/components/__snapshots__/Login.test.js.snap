// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> renders the Login component correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login
    auth={
      Object {
        "addMatchers": [Function],
        "advanceTimersByTime": [Function],
        "autoMockOff": [Function],
        "autoMockOn": [Function],
        "clearAllMocks": [Function],
        "clearAllTimers": [Function],
        "deepUnmock": [Function],
        "disableAutomock": [Function],
        "doMock": [Function],
        "dontMock": [Function],
        "enableAutomock": [Function],
        "fn": [Function],
        "genMockFromModule": [Function],
        "isMockFunction": [Function],
        "mock": [Function],
        "requireActual": [Function],
        "requireMock": [Function],
        "resetAllMocks": [Function],
        "resetModuleRegistry": [Function],
        "resetModules": [Function],
        "restoreAllMocks": [Function],
        "retryTimes": [Function],
        "runAllImmediates": [Function],
        "runAllTicks": [Function],
        "runAllTimers": [Function],
        "runOnlyPendingTimers": [Function],
        "runTimersToTime": [Function],
        "setMock": [Function],
        "setTimeout": [Function],
        "spyOn": [Function],
        "unmock": [Function],
        "useFakeTimers": [Function],
        "useRealTimers": [Function],
      }
    }
    errors={
      Object {
        "addMatchers": [Function],
        "advanceTimersByTime": [Function],
        "autoMockOff": [Function],
        "autoMockOn": [Function],
        "clearAllMocks": [Function],
        "clearAllTimers": [Function],
        "deepUnmock": [Function],
        "disableAutomock": [Function],
        "doMock": [Function],
        "dontMock": [Function],
        "enableAutomock": [Function],
        "fn": [Function],
        "genMockFromModule": [Function],
        "isMockFunction": [Function],
        "mock": [Function],
        "requireActual": [Function],
        "requireMock": [Function],
        "resetAllMocks": [Function],
        "resetModuleRegistry": [Function],
        "resetModules": [Function],
        "restoreAllMocks": [Function],
        "retryTimes": [Function],
        "runAllImmediates": [Function],
        "runAllTicks": [Function],
        "runAllTimers": [Function],
        "runOnlyPendingTimers": [Function],
        "runTimersToTime": [Function],
        "setMock": [Function],
        "setTimeout": [Function],
        "spyOn": [Function],
        "unmock": [Function],
        "useFakeTimers": [Function],
        "useRealTimers": [Function],
      }
    }
    history={
      Object {
        "push": [MockFunction],
      }
    }
    signIn={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="loginbox"
      >
        <img
          className="avatar"
          src="https://store--manager.herokuapp.com/assets/uploads/users/default-avatar.png"
        />
        <h2
          className="text-center"
        >
          Login to access store manager
        </h2>
        <form
          onSubmit={[Function]}
        >
          <p>
            Email
          </p>
          <input
            id="loginusername"
            name="email"
            onChange={[Function]}
            placeholder="Enter Email"
            type="text"
          />
          <p>
            Password
          </p>
          <input
            id="loginpassword"
            name="password"
            onChange={[Function]}
            placeholder="Enter Password"
            type="password"
          />
          <input
            id="loginsubmit"
            name="submit"
            type="submit"
            value="Login"
          />
        </form>
      </div>,
      "id": "loginbody",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <img
            className="avatar"
            src="https://store--manager.herokuapp.com/assets/uploads/users/default-avatar.png"
          />,
          <h2
            className="text-center"
          >
            Login to access store manager
          </h2>,
          <form
            onSubmit={[Function]}
          >
            <p>
              Email
            </p>
            <input
              id="loginusername"
              name="email"
              onChange={[Function]}
              placeholder="Enter Email"
              type="text"
            />
            <p>
              Password
            </p>
            <input
              id="loginpassword"
              name="password"
              onChange={[Function]}
              placeholder="Enter Password"
              type="password"
            />
            <input
              id="loginsubmit"
              name="submit"
              type="submit"
              value="Login"
            />
          </form>,
        ],
        "className": "loginbox",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "avatar",
            "src": "https://store--manager.herokuapp.com/assets/uploads/users/default-avatar.png",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Login to access store manager",
            "className": "text-center",
          },
          "ref": null,
          "rendered": "Login to access store manager",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <p>
                Email
              </p>,
              <input
                id="loginusername"
                name="email"
                onChange={[Function]}
                placeholder="Enter Email"
                type="text"
              />,
              <p>
                Password
              </p>,
              <input
                id="loginpassword"
                name="password"
                onChange={[Function]}
                placeholder="Enter Password"
                type="password"
              />,
              <input
                id="loginsubmit"
                name="submit"
                type="submit"
                value="Login"
              />,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Email",
              },
              "ref": null,
              "rendered": "Email",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "id": "loginusername",
                "name": "email",
                "onChange": [Function],
                "placeholder": "Enter Email",
                "type": "text",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Password",
              },
              "ref": null,
              "rendered": "Password",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "id": "loginpassword",
                "name": "password",
                "onChange": [Function],
                "placeholder": "Enter Password",
                "type": "password",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "id": "loginsubmit",
                "name": "submit",
                "type": "submit",
                "value": "Login",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
          ],
          "type": "form",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="loginbox"
        >
          <img
            className="avatar"
            src="https://store--manager.herokuapp.com/assets/uploads/users/default-avatar.png"
          />
          <h2
            className="text-center"
          >
            Login to access store manager
          </h2>
          <form
            onSubmit={[Function]}
          >
            <p>
              Email
            </p>
            <input
              id="loginusername"
              name="email"
              onChange={[Function]}
              placeholder="Enter Email"
              type="text"
            />
            <p>
              Password
            </p>
            <input
              id="loginpassword"
              name="password"
              onChange={[Function]}
              placeholder="Enter Password"
              type="password"
            />
            <input
              id="loginsubmit"
              name="submit"
              type="submit"
              value="Login"
            />
          </form>
        </div>,
        "id": "loginbody",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <img
              className="avatar"
              src="https://store--manager.herokuapp.com/assets/uploads/users/default-avatar.png"
            />,
            <h2
              className="text-center"
            >
              Login to access store manager
            </h2>,
            <form
              onSubmit={[Function]}
            >
              <p>
                Email
              </p>
              <input
                id="loginusername"
                name="email"
                onChange={[Function]}
                placeholder="Enter Email"
                type="text"
              />
              <p>
                Password
              </p>
              <input
                id="loginpassword"
                name="password"
                onChange={[Function]}
                placeholder="Enter Password"
                type="password"
              />
              <input
                id="loginsubmit"
                name="submit"
                type="submit"
                value="Login"
              />
            </form>,
          ],
          "className": "loginbox",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "avatar",
              "src": "https://store--manager.herokuapp.com/assets/uploads/users/default-avatar.png",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Login to access store manager",
              "className": "text-center",
            },
            "ref": null,
            "rendered": "Login to access store manager",
            "type": "h2",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <p>
                  Email
                </p>,
                <input
                  id="loginusername"
                  name="email"
                  onChange={[Function]}
                  placeholder="Enter Email"
                  type="text"
                />,
                <p>
                  Password
                </p>,
                <input
                  id="loginpassword"
                  name="password"
                  onChange={[Function]}
                  placeholder="Enter Password"
                  type="password"
                />,
                <input
                  id="loginsubmit"
                  name="submit"
                  type="submit"
                  value="Login"
                />,
              ],
              "onSubmit": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Email",
                },
                "ref": null,
                "rendered": "Email",
                "type": "p",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "id": "loginusername",
                  "name": "email",
                  "onChange": [Function],
                  "placeholder": "Enter Email",
                  "type": "text",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Password",
                },
                "ref": null,
                "rendered": "Password",
                "type": "p",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "id": "loginpassword",
                  "name": "password",
                  "onChange": [Function],
                  "placeholder": "Enter Password",
                  "type": "password",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "id": "loginsubmit",
                  "name": "submit",
                  "type": "submit",
                  "value": "Login",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
            ],
            "type": "form",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
